include/geos/operation/cluster/AbstractClusterFinder.h

```diff
@@ -0,0 +1,136 @@
+/**********************************************************************
+ *
+ * GEOS - Geometry Engine Open Source
+ * http://geos.osgeo.org
+ *
+ * Copyright (C) 2020-2021 Daniel Baston
+ *
+ * This is free software; you can redistribute and/or modify it under
+ * the terms of the GNU Lesser General Public Licence as published
+ * by the Free Software Foundation.
+ * See the COPYING file for more information.
+ *
+ **********************************************************************/
+
+#ifndef GEOS_OPERATION_CLUSTER_ABSTRACTCLUSTERFINDER
+#define GEOS_OPERATION_CLUSTER_ABSTRACTCLUSTERFINDER
+
+#include <functional>
+#include <memory>
+#include <vector>
+
+#include <geos/export.h>
+#include <geos/index/strtree/TemplateSTRtree.h>
+#include <geos/operation/cluster/Clusters.h>
+
+// Forward declarations
+namespace geos {
+namespace geom {
+    class Envelope;
+    class Geometry;
+}
+namespace operation {
+namespace cluster {
+    class UnionFind;
+}
+}
+}
+
+namespace geos {
+namespace operation {
+namespace cluster {
+
+/** AbstractClusterFinder defines an interface for bottom-up clustering algorithms,
+ * where spatial index queries can be used to identify geometries that should be
+ * clustered together.
+ */
+class GEOS_DLL AbstractClusterFinder {
+
+public:
+    /**
+     * Cluster the provided geometries, returning an object that provides access
+     * to the components of each cluster.
+     *
+     * @param g A vector of geometries to cluster
+     */
+    Clusters cluster(const std::vector<const geom::Geometry*>& g);
+
+    /**
+     * Cluster the components of the provided geometry, returning a vector of clusters.
+     * This function will take ownership of the provided geometry. Any components that
+     * are included in a cluster will be returned. Components that are not included in
+     * any cluster will be destroyed.
```

Comment:
Doesn't this semantic need to be documented in the C API?  Or is it inferred?

Reply:
It can be inferred from the fact that we're passing a non-const pointer, but it can also be documented. But I am leaning towards removing the C API until I have a better understanding of if/how it would be used from shapely et al.